### PR TITLE
Refactor operator estimation and replace measure_observables

### PR DESCRIPTION
**Pull request by kylegulshen**
This is the first step in refactoring this repo to use a standard workflow and move away from dataframes.

`operator_estimation.py` originated in this repo but was copied over to pyquil and modified extensively. The version of the file in fb/master is outdated, and only the pyquil version is currently ever used within forest benchmarking. Commit 504f032 replaces the outdated version with the latest version from pyquil. Subsequently commit 31aa7aa reflects my changes to the pyquil version along with the updates to the tests. Looking at the diff between these commits will be more instructive, and in particular looking at the changes to the tests should provide a general sense of the change in usage. I've also included a notebook with outputs preserved for convenient inspection of usage. 

Note that some things were simply renamed rather than substantively changed. The main update is to decompose measure_observables into several steps, each of which works with lists of programs.

### CODE REVIEW

File: forest/benchmarking/compilation.py [hunk 1]
```diff
@@ -57,6 +57,17 @@ def _RY(angle, q):
     return p
 
 
+def _RX(angle, q):
+    """
+    A RX in terms of RX(+-pi/2) and RZ(theta)
```
Comment (blakejohnson) [inline forest/benchmarking/compilation.py]: The docstring doesn't match the generated `Program`. I think you intended to swap `RZ` and `_RY` below?

File: forest/benchmarking/operator_estimation.py [hunk 2]
```diff
@@ -1,557 +1,1210 @@
...
+    assert bitarray.shape[1] == len(qubits), 'qubits should label each column of the bitarray'
+
+    # Pick columns corresponding to qubits with a non-identity out_operation
+    obs_strings = bitarray[:, idxs]
+    # Transform bits to eigenvalues; ie (+1, -1)
+    my_obs_strings = 1 - 2 * obs_strings
+    # Multiply row-wise to get operator values.
+    obs_vals = np.prod(my_obs_strings, axis=1)
+
+    if use_beta_dist_unbiased_prior:
```
Comment (joshcombes) [inline forest/benchmarking/operator_estimation.py]: So I thought about this for a bit (the reasoning also applies to the frequentist version below.

suppose the true value is  `p=1e-N` for some positive N. We will roughly need num_shots proportional to 1/p in order to get the correct accuracy.

File: forest/benchmarking/operator_estimation.py [hunk 3]
```diff
@@ -1,557 +1,1210 @@
...
+    assert method in allowed_methods, f"'method' should be one of {allowed_methods}."
+    if method == 'greedy':
+        return group_experiments_greedy(experiments)
+    elif method == 'clique-removal':
+        return group_experiments_clique_removal(experiments)
+
+
+@dataclass(frozen=True)
+class ExperimentResult:
+    """An expectation and standard deviation for the measurement of one experiment setting
```
Comment (joshcombes) [inline forest/benchmarking/operator_estimation.py]: in an observables experiment.

File: forest/benchmarking/operator_estimation.py [hunk 4]
```diff
@@ -1,557 +1,1210 @@
...
+            if oneq_state.qubit in mapping:
+                if mapping[oneq_state.qubit] != oneq_state:
+                    return None
+            else:
+                mapping[oneq_state.qubit] = oneq_state
+    return TensorProductState(list(mapping.values()))
 
 
-def commuting_sets_by_zbasis(pauli_sums):
+def _max_tpb_overlap(tomo_expt: ObservablesExperiment):
```
Comment (joshcombes) [inline forest/benchmarking/operator_estimation.py]: `tomo_expt` -> `obs_expt` ?

File: forest/benchmarking/operator_estimation.py [hunk 5]
```diff
@@ -1,557 +1,1210 @@
...
+        and can be run together
+    :return: dictionary keyed with ExperimentSetting (specifying a tpb), and with each value being a
+            list of ExperimentSettings (diagonal in that tpb)
     """
+    # initialize empty dictionary
     diagonal_sets = {}
-    for term in pauli_sums:
-        diagonal_sets = _max_key_overlap(term, diagonal_sets)
+    # loop through ExperimentSettings of the ObservablesExperiment
+    for expt_setting in tomo_expt:
```
Comment (joshcombes) [inline forest/benchmarking/operator_estimation.py]: `tomo_expt` -> `obs_expt` ?

File: forest/benchmarking/compilation.py [hunk 6]
```diff
@@ -57,6 +57,18 @@ def _RY(angle, q):
     return p
 
 
+def _RX(angle, q):
+    """
+    A RX in terms of RZ(+-pi/2) and _RY(theta), which is itself decomposed
+    into native gates RX(+-pi/2) and RZ(theta) above.
```
Comment (blakejohnson) [inline forest/benchmarking/compilation.py]: Ok, you've made the docstring consistent with the program, but this way of compiling an RX doesn't make much sense to me, unless you are trying to re-use code from the definition of `_RY`. This is one case where that code re-use instinct works counter to code readability, and so you should instead just write a proper definition of RX in terms of Zs and RX(pi/2)s.

Comment (kylegulshen) [reply forest/benchmarking/compilation.py]: I'm still not sure if I executed your suggestion correctly--if you mean that the sequence of gates itself should be different, then I'm not sure what sequence is simpler. The sequence here matches the output of the compiler.

**Comment by mpharrigan** [comment]
wait, why are you copying all of this in from pyquil? Why not just use the pyquil version and contribute changes there, "upstream"? Or layer your changes on top of it

**Comment by joshcombes** [comment]
Hey Matt, based on some requirements from the lower level of the stack we (well Kyle) are refactoring `operator_estimation`. The idea would be to port it back to pyQuil. But at the moment we need to rapidly iterate the code and that is hard to do in pyQuil.

**Pull request merged**